internal/argocd/revision.go

```diff
@@ -2,107 +2,224 @@ package argocd
 
 import (
 	"context"
+	"errors"
 	"fmt"
 
 	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	kargoapi "github.com/akuity/kargo/api/v1alpha1"
 	argocd "github.com/akuity/kargo/internal/controller/argocd/api/v1alpha1"
 	"github.com/akuity/kargo/internal/controller/freight"
+	"github.com/akuity/kargo/internal/logging"
 )
 
+var NO_REVISIONS = []string{}
+
+func GetIntersection(revisions []string, desired []string) []string {
+	var intersection []string
+	hash := make(map[string]bool)
+	for _, e := range revisions {
+		hash[e] = true
+	}
+	for _, e := range desired {
+		// If elements present in the hashmap then append intersection list.
+		if hash[e] {
+			intersection = append(intersection, e)
+		}
+	}
+	return intersection
+}
+
 // GetDesiredRevision returns the desired revision for the given
 // v1alpha1.Application. If that cannot be determined, an empty string is
 // returned.
-func GetDesiredRevision(
+func GetDesiredRevisions(
 	ctx context.Context,
 	cl client.Client,
 	stage *kargoapi.Stage,
 	update *kargoapi.ArgoCDAppUpdate,
 	app *argocd.Application,
 	frght []kargoapi.FreightReference,
-) (string, error) {
+) ([]string, error) {
+
+	logger := logging.LoggerFromContext(ctx)
+	if app == nil {
+		err := errors.New("Application is nil")
+		logger.Error(err, "Cannot determine desired revision for application, bailing.")
+		return NO_REVISIONS, nil
+	}
+
+	appLogger := logger.WithValues("application", app.Name, "namespace", app.Namespace)
+	appLogger.Debug("Getting desired revision for application", "app", app, "spec", app.Spec)
+
 	// Note that frght was provided as an argument instead of being plucked
 	// directly from stage.Status, because this gives us the flexibility to use
 	// this function for finding the revision to sync to either in the context of
 	// a health check (current freight) or in the context of a promotion (new
 	// freight).
+
+	// An application may have one or more sources.
+	if !app.IsMultisource() {
+		if app.Spec.Source == nil {
+			err := errors.New("Single-source application source is nil")
+			appLogger.Error(err, "Cannot determine desired revision for application, bailing.")
+			return NO_REVISIONS, nil
+		}
+
+		appLogger.Debug("Application source is not nil, checking.", "source", app.Spec.Source)
+		desiredRevision, err := getRevisionFromSource(ctx, cl, stage, update, app.Spec.Source, frght)
+
+		if err != nil {
+			return NO_REVISIONS, err
+		}
+
+		if desiredRevision != "" {
+			return []string{desiredRevision}, nil
+		}
+	}
+
+	// An application may have more than one source that points to the same Git repository,
+	// eg. a Helm and a vanilla manifest source.
+	// In that situation it doesn't matter which source's target revision is returned as ArgoCD does not support
+	// different target revisions for sources targeting the same repository.
+	var revisions []string
+	for i := range app.Spec.Sources {
+		s := &app.Spec.Sources[i]
```

Comment:
It is generally more idiomatic do do this:

```suggestion
	for _, s := range app.Spec.Sources {
```

If you've seen other places in the code where we used the index, there was probably a specific reason for it, which is often that the slice is being modified in place as we iterate. That isn't applicable here, so you can get away with the more idiomatic approach.

This applies in several other spots as well.

Reply:
I found plenty of examples for both so I picked one, but I don't have a preference.

Reply:
Ah, the linter just reminded me why: "Implicit memory aliasing in for loop."

```
internal/argocd/revision.go:85:73: G601: Implicit memory aliasing in for loop. (gosec)
                desiredRevision, err := getRevisionFromSource(ctx, cl, stage, update, &s, frght)
```

I'm going to keep my original unless you have a better alternative.

Reply:
In that case, we might typically do this:

```
for _, src := range app.Spec.Sources {
  s := src
```

iirc, Go 1.21+ automatically compensate  for implicit memory aliasing, but the linter hasn't got the memo yet. 😉 

Just leave it be.